controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/UpgraderTest.java

```diff
@@ -128,15 +136,24 @@ public void testUpgrading() {
         tester.completeUpgrade(default2, version, "default");
 
         tester.updateVersionStatus(version);
-        assertEquals("Not enough evidence to mark this neither broken nor high",
+        assertEquals("Not enough evidence to mark this as neither broken nor high",
                      VespaVersion.Confidence.normal, tester.controller().versionStatus().systemVersion().get().confidence());
-        tester.upgrader().maintain();
+
         assertEquals("Upgrade with error should retry", 1, tester.buildSystem().jobs().size());
 
+        // Finish previous run, with exhausted retry.
+        tester.clock().advance(Duration.ofHours(1));
+        tester.notifyJobCompletion(DeploymentJobs.JobType.stagingTest, default0, false);
+
         // --- Failing application is repaired by changing the application, causing confidence to move above 'high' threshold
+        // TODO: NO: Change only carries a single change at a time, even though there are two in this case, leaving the version at 5.1 for this application.
```

Comment:
This is something to consider. I had to add the below to make the test (and logic) right. Perhaps we should rather use a Target with Optional .version() and .revision() methods, instead of Change with subclasses? I think that's both cleaner, it allows modelling simultaneous changes — like we apparently support now — and the current Change is used mostly as a Target, anway.

Reply:
Not sure what you mean by "No" here, did you disallow this? It need to be allowed.
Change carries a single change, but it only *decides* the version, it does not decide the revision.

Regarding Change: I want to get rid of it too.
We should definitely have an optional target version, but it may be better to just infer the revision from the observed job status on demand as that will match what is actually happening.

Reply:
I mean that in the current implementation, the ongoing upgrade is overridden, since the application will get a new deploying(), which is an ApplicationChange, from the new component job run. 
Thus, the first pass through the pipeline upgrades the revision, and a second pass is needed for the version to reach production.

I can't say exactly why the test didn't fail before, but the isRunning() was not well behaved for the time-less testing tools (DeploymentTester et al), and I suspect that's why it wasn't caught.

Reply:
No, an application change will not overwrite an ongoing version change:

            // Handle successful starting and ending
            if (report.success()) {
                if (order.givesApplicationChange(report.jobType())) {
                    if (acceptNewRevisionNow(application)) {
                        // Set this as the change we are doing, unless we are already pushing a platform change
                        if ( ! ( application.deploying().isPresent() && 
                                 (application.deploying().get() instanceof Change.VersionChange)))
                            application = application.withDeploying(Optional.of(Change.ApplicationChange.unknown()));
                    }
                } 
                ...
            }

Anyway, this confusion just illustrates that we should change this model as suggested above.

Reply:
Argh, another condition, hidden away somewhere! I missed that one.

Reply:
So what happens, then, when there is a failing upgrade? The new application revision is carried, in stealth, until it catches the failing version upgrade?

Reply:
Also, reinserting the condition I removed (which you posted above), made the test succeed without the extra pass.

Reply:
The deployment will still be counted as a version change even though it also changes the revision, yes. It's not that stealthy though, since you can see it in what the jobs are doing. Which is why I suggest we just remove the ApplicationChange thing - it was the best we had before we tracked the jobs like this but now it is just a sometimes-incorrect summary of that information.